pkg/reconciler/committer/committer.go

```diff
@@ -0,0 +1,121 @@
+	Spec              Sp `json:"spec"`
+	Status            St `json:"status,omitempty"`
+}
+
+// Patcher is just the Patch API with a generic to keep use sites type safe
+type Patcher[R any] interface {
+	Patch(ctx context.Context, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (R, error)
+}
+
+// NewCommitter returns a function that can patch instances of R based on spec or status changes.
+func NewCommitter[R any, Sp any, St any](patcher Patcher[R]) func(context.Context, *Resource[Sp, St], *Resource[Sp, St]) error {
+	focusType := fmt.Sprintf("%T", *new(R))
+	return func(ctx context.Context, old, obj *Resource[Sp, St]) error {
+		logger := klog.FromContext(ctx)
+
+		objectMetaChanged := !equality.Semantic.DeepEqual(old.ObjectMeta, obj.ObjectMeta)
+		specChanged := !equality.Semantic.DeepEqual(old.Spec, obj.Spec)
+		statusChanged := !equality.Semantic.DeepEqual(old.Status, obj.Status)
+
+		specOrObjectMetaChanged := specChanged || objectMetaChanged
+
+		if !specOrObjectMetaChanged && !statusChanged {
+			return nil
+		}
+
+		// forPatch ensures that only the spec/objectMeta fields will be changed
+		// or the status field but never both at the same time.
+		forPatch := func(r *Resource[Sp, St]) *Resource[Sp, St] {
+			var ret Resource[Sp, St]
+			if specOrObjectMetaChanged {
+				ret.ObjectMeta = r.ObjectMeta
+				ret.Spec = r.Spec
+			} else {
+				ret.Status = r.Status
+			}
+			return &ret
+		}
+
+		clusterName := logicalcluster.From(old)
+		name := old.Name
+
+		oldForPatch := forPatch(old)
+		// to ensure they appear in the patch as preconditions
+		oldForPatch.UID = ""
+		oldForPatch.ResourceVersion = ""
+
+		oldData, err := json.Marshal(oldForPatch)
+		if err != nil {
+			return fmt.Errorf("failed to Marshal old data for %s %s|%s: %w", focusType, clusterName, name, err)
+		}
+
+		newForPatch := forPatch(obj)
+		// to ensure they appear in the patch as preconditions
+		newForPatch.UID = old.UID
+		newForPatch.ResourceVersion = old.ResourceVersion
+
+		newData, err := json.Marshal(newForPatch)
+		if err != nil {
+			return fmt.Errorf("failed to Marshal new data for %s %s|%s: %w", focusType, clusterName, name, err)
+		}
+
+		patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
+		if err != nil {
+			return fmt.Errorf("failed to create patch for %s %s|%s: %w", focusType, clusterName, name, err)
+		}
+
+		var subresources []string
+		if statusChanged {
+			subresources = []string{"status"}
+		}
+
+		logger.V(2).Info(fmt.Sprintf("patching %s", focusType), "patch", string(patchBytes))
+		_, err = patcher.Patch(logicalcluster.WithCluster(ctx, clusterName), obj.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, subresources...)
+		if err != nil {
+			return fmt.Errorf("failed to patch %s %s|%s: %w", focusType, clusterName, name, err)
+		}
+
+		// Simultaneous updates of spec and status are never allowed.
+		if specOrObjectMetaChanged && statusChanged {
```

Comment:
@stevekuznetsov Shouldn't this go near the top?

Reply:
yes, that should be as early as possible